compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs

```diff
@@ -223,10 +223,17 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             tcx.parent_hir_node(self_ty.hir_id),
             hir::Node::Ty(hir::Ty { kind: hir::TyKind::Ref(..), .. })
         );
+        let is_non_trait_object = |ty: &'tcx hir::Ty<'_>| {
```

Comment:
This feels unfortunately somewhat ad-hoc, since any place we want to inspect the `hir::TyKind`, like for *any* diagnostics, we'll need to call the `get_future_inner_return_ty`.